federation/pkg/federation-controller/service/dns/dns.go

```diff
@@ -0,0 +1,431 @@
+				}
+			}
+		}
+	}
+
+	if err := changeSet.Apply(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/* ensureDnsRecords ensures (idempotently, and with minimum mutations) that all of the DNS records for a service are correct.
+This should be called every time the state of a service might have changed
+(either w.r.t. it's loadbalancer address, or if the number of healthy backend endpoints for that service transitioned from zero to non-zero
+(or vice verse).  Only shards of the service which have both a loadbalancer ingress IP address or hostname AND at least one healthy backend endpoint
+are included in DNS records for that service (at all of zone, region and global levels). All other addresses are removed.  Also, if no shards exist
+in the zone or region of the cluster, a CNAME reference to the next higher level is ensured to exist. */
+func (d *ServiceDNSController) ensureDnsRecords(service *v1.Service) {
+	// Quinton: Pseudocode....
+	// See https://github.com/kubernetes/kubernetes/pull/25107#issuecomment-218026648
+	// For each service we need the following DNS names:
+	// mysvc.myns.myfed.svc.z1.r1.mydomain.com  (for zone z1 in region r1)
+	//         - an A record to IP address of specific shard in that zone (if that shard exists and has healthy endpoints)
+	//         - OR a CNAME record to the next level up, i.e. mysvc.myns.myfed.svc.r1.mydomain.com  (if a healthy shard does not exist in zone z1)
+	// mysvc.myns.myfed.svc.r1.federation
+	//         - a set of A records to IP addresses of all healthy shards in region r1, if one or more of these exist
+	//         - OR a CNAME record to the next level up, i.e. mysvc.myns.myfed.svc.mydomain.com (if no healthy shards exist in region r1)
+	// mysvc.myns.myfed.svc.federation
+	//         - a set of A records to IP addresses of all healthy shards in all regions, if one or more of these exist.
+	//         - no record (NXRECORD response) if no healthy shards exist in any regions)
+
+	key := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}.String()
+
+	dnsNamePrefix := service.Name + "." + service.Namespace + "." + d.federationName + ".svc"
+
+	ingress := &fed.FederatedServiceIngress{}
+	if service.DeletionTimestamp != nil {
+		// if service is about to  be deleted, remove all ingress endpoints associated with the service
+		glog.V(2).Infof("Deleting federation dns records for %s service", key)
+	} else {
+		// otherwise, parse the ingress from service annotations.
+		var err error
+		ingress, err = si.ParseFederationServiceIngresses(service)
+		if err != nil {
+			glog.Errorf("Error in parsing lb ingress for service %s/%s: %v", service.Namespace, service.Name, err)
+			return
+		}
+	}
+
+	endpointsMap := make(map[string][]string)
+	globalEndpoints := []string{}
+	for regionName, region := range ingress.Endpoints {
+		regionEndpoints := []string{}
+		for zoneName, zone := range region {
```

Comment:
Similar to above comment, 'endpoints' rather than 'zone' here?  The current code is confusing, at least to me.